include/gz/sim/components/Environment.hh

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GZ_SIM_ENVIRONMENT_HH_
+#define GZ_SIM_ENVIRONMENT_HH_
+
+#include <memory>
+#include <string>
+
+#include <gz/common/DataFrame.hh>
+#include <gz/math/SphericalCoordinates.hh>
+#include <gz/math/TimeVaryingVolumetricGrid.hh>
+
+#include <gz/sim/components/Factory.hh>
+#include <gz/sim/components/Component.hh>
+
+namespace gz
+{
+namespace sim
+{
+// Inline bracket to help doxygen filtering.
+inline namespace GZ_SIM_VERSION_NAMESPACE {
+namespace components
+{
+  /// \brief Environment data across time and space. This is useful to
+  /// introduce physical quantities that may be of interest even if not
+  /// modelled in simulation. 
+  struct EnvironmentalData
+  {
+    using T = math::InMemoryTimeVaryingVolumetricGrid<double>;
+    using FrameT = common::DataFrame<std::string, T>;
+    using ReferenceT = math::SphericalCoordinates::CoordinateType;
+
+    /// \brief Instantiate environmental data.
+    ///
+    /// An std::make_shared equivalent that ensures
+    /// dynamically loaded call sites use a template
+    /// instantiation that is guaranteed to outlive
+    /// them.
+    static std::shared_ptr<EnvironmentalData>
+    MakeShared(FrameT _frame, ReferenceT _reference);
+
+    /// \brief Environmental data frame.
+    FrameT frame;
+
+    /// \brief Spatial reference for data coordinates. 
+    ReferenceT reference;
```

Comment:
I assume `CoordinateType::SPHERICAL` means we use lat-long. On the other hand `LOCAL` frame makes no sense here. `GLOBAL` and `ECEF` also makes sense, however I don't know if we want to support `ECEF` right away?

Reply:
@arjo129 I'll start by saying that I chose to use `gz::math::SphericalCoordinates` as a reference only because it was convenient. Regarding coordinate system types, what I gather from https://github.com/gazebosim/gz-math/blob/6de6120e4d610607b71cb03242862aa07d2582bb/src/SphericalCoordinates.cc#L464-L473 is that `LOCAL` is a rotation away (that specified by heading offsets in spherical coordinates) from `GLOBAL`. In any case, no code should have to deal with these coordinate systems explicitly. It's just how the data coordinates are expressed. 

> I assume `CoordinateType::SPHERICAL` means we use lat-long.

It means the data uses lat-lon coordinates. Code may as well be using global cartesian coordinates. As long as the conversion gets carried out before looking up the frame (see https://github.com/osrf/lrauv/pull/241/files#diff-294ae3632db44c65a140d81aa17a033ddf905909f1093087116e231742ba4c06R1690-R1695 for an example), it'll be fine. Same for `ECEF`.